File: backend/api/svm.py
from typing import List
from fastapi import APIRouter, HTTPException

from .models import (
    ParameterInfo,
    SVMTrainRequest,
    SVMTrainResponse,
    SVMVisualisationRequest,
    SVMVisualisationResponse,
    SVMStepRequest,
    SVMStepResponse,
    SVMPredictRequest,
    SVMPredictResponse
)
from services.svm_service import svm_service

router = APIRouter()

@router.get("/params", response_model=List[ParameterInfo])
async def get_parameters() -> List[ParameterInfo]:
    """Get the parameters for SVM Classification."""
    return await svm_service.get_parameters()

@router.post("/visualise", response_model=SVMVisualisationResponse)
async def visualise(
    request: SVMVisualisationRequest,
) -> SVMVisualisationResponse:
    try:
        result = await svm_service.visualise(
            parameters=request.parameters,
            dataset_param=request.dataset,
        )
        return SVMVisualisationResponse(**result)
    except Exception as e:
        raise HTTPException(status_code=400, detail=str(e))

@router.post("/train", response_model=SVMTrainResponse)
async def train(
    request: SVMTrainRequest,
) -> SVMTrainResponse:
    try:
        result = await svm_service.train(
            parameters=request.parameters,
            dataset_param=request.dataset,
        )
        return SVMTrainResponse(**result)
    except Exception as e:
        raise HTTPException(status_code=400, detail=str(e))

@router.post("/step", response_model=SVMStepResponse)
async def step(
    request: SVMStepRequest,
) -> SVMStepResponse:
    try:
        result = await svm_service.step(
            current_w1=request.current_w1,
            current_w2=request.current_w2,
            current_b=request.current_b,
            learning_rate=request.learning_rate,
            parameters=request.parameters,
            dataset_param=request.dataset,
        )
        return SVMStepResponse(**result)
    except Exception as e:
        raise HTTPException(status_code=400, detail=str(e))

@router.post("/predict", response_model=SVMPredictResponse)
async def predict(
    request: SVMPredictRequest,
) -> SVMPredictResponse:
    try:
        result = await svm_service.predict(
            w1=request.w1,
            w2=request.w2,
            b=request.b,
            parameters=request.parameters,
            dataset_param=request.dataset,
        )
        return SVMPredictResponse(**result)
    except Exception as e:
        raise HTTPException(status_code=400, detail=str(e))

File: frontend/src/api/svm.ts
/**
 * API methods for Support Vector Machine (SVM) Classification
 */
import { API_BASE_URL as BASE_URL } from "@/api/config";
import type { ParameterInfo } from "@/api/types";
import type { components } from "@/types/api";

export type SVMTrainRequest = components["schemas"]["SVMTrainRequest"];
export type SVMTrainResponse = components["schemas"]["SVMTrainResponse"];
export type SVMPredictRequest = components["schemas"]["SVMPredictRequest"];
export type SVMPredictResponse = components["schemas"]["SVMPredictResponse"];
export type SVMStepRequest = components["schemas"]["SVMStepRequest"];
export type SVMStepResponse = components["schemas"]["SVMStepResponse"];
export type SVMVisualisationRequest = components["schemas"]["SVMVisualisationRequest"];
export type SVMVisualisationResponse = components["schemas"]["SVMVisualisationResponse"];

// For internal frontend use if needed
export interface SVMPoint {
    id: number;
    x: [number, number];
    y: number; // class label
    margin_distance?: number;
    is_support_vector?: boolean;
}

const API_BASE_URL = `${BASE_URL}/api/svm`;

// --- API Calls ---

export const getParameters = async (): Promise<ParameterInfo[]> => {
    const response = await fetch(`${API_BASE_URL}/params`);
    if (!response.ok) throw new Error(`HTTP error! status: ${response.status}`);
    return response.json();
};

export const trainSVM = async (
    request: Partial<SVMTrainRequest> = {}
): Promise<SVMTrainResponse> => {
    const response = await fetch(`${API_BASE_URL}/train`, {
        method: "POST",
        headers: { "Content-Type": "application/json" },
        body: JSON.stringify(request),
    });
    const data: SVMTrainResponse = await response.json();
    if (!response.ok || !data.success) {
        throw new Error((data as any).detail || `HTTP error! status: ${response.status}`);
    }
    return data;
};

export const getSVMPrediction = async (
    request: Partial<SVMPredictRequest> = {}
): Promise<SVMPredictResponse> => {
    const response = await fetch(`${API_BASE_URL}/predict`, {
        method: "POST",
        headers: { "Content-Type": "application/json" },
        body: JSON.stringify(request),
    });
    const data: SVMPredictResponse = await response.json();
    if (!response.ok || !data.success) {
        throw new Error((data as any).detail || `HTTP error! status: ${response.status}`);
    }
    return data;
};

export const stepSVM = async (
    request: Partial<SVMStepRequest>
): Promise<SVMStepResponse> => {
    const response = await fetch(`${API_BASE_URL}/step`, {
        method: "POST",
        headers: { "Content-Type": "application/json" },
        body: JSON.stringify(request),
    });
    const data: SVMStepResponse = await response.json();
    if (!response.ok || !data.success) {
        throw new Error((data as any).detail || `HTTP error! status: ${response.status}`);
    }
    return data;
};
